js/src/components/adaptive-form/adaptive-form-context.js

```diff
@@ -0,0 +1,42 @@
+/**
+ * External dependencies
+ */
+import { createContext, useContext } from '@wordpress/element';
+
+/**
+ * @typedef {Object} InputProps
+ * @property {*} value Form value.
+ * @property {boolean} checked Form value converted to boolean.
+ * @property {*} selected Form value.
+ * @property {(value: Event<HTMLInputElement> | *) => void} onChange Function to handle onChange event.
+ * @property {() => void} onBlur Function to handle onBlur event.
+ * @property {'has-error'|undefined} className 'has-error' if the form value is invalid and marked as touched. `undefined` otherwise.
+ * @property {string|undefined|null} help The corresponding value in form `errors` if the form value is marked as touched. `null` otherwise.
+ */
+
+/**
+ * @typedef {Object} AdaptiveFormContext
+ * @property {Object} values Form values.
+ * @property {Object} errors Object with key-value pairs representing errors for form values. Empty object if no errors.
+ * @property {Object} touched Object with key-value pairs representing form values is touched.
+ * @property {boolean} isValidForm `true` if form values pass the validation.
+ * @property {boolean} isDirty `true` after any of the form values is modified.
+ * @property {(name: string, value: *) => void} setValue Function to set a form value.
+ * @property {(name: string) => InputProps} getInputProps Function to get the corresponding input props by the name of a form value.
+ * @property {() => Promise<Object>} handleSubmit Function to trigger form submission.
+ * @property {(initialValues: Object) => void} resetForm Function to reset form with given initial values.
+ * @property {(currentTouched: Object) => void} setTouched Function to update the `touched` state.
+ */
+
+export const AdaptiveFormContext = createContext( {} );
+
+/**
+ * AdaptiveForm's context hook.
+ *
+ * @return {AdaptiveFormContext} AdaptiveForm's context.
+ */
+export function useAdaptiveFormContext() {
+	const adaptiveFormContext = useContext( AdaptiveFormContext );
+
+	return adaptiveFormContext;
+}
```

Comment:
I agree with the use of context for this case, but I found this hook `useAdaptiveFormContext` a bit redundant tho.

Why not to have a context folder with the contexts (this one for now) and just call useContext() in the consumer side?

Reply:
> I agree with the use of context for this case, but I found this hook useAdaptiveFormContext a bit redundant tho.

Since there will be many form-related components that will use this context, there will be many such code fragments repeated across them.

```js
import { useContext } from '@wordpress/element';
import { AdaptiveFormContext } from '.~/components/adaptive-form';

function InputComponent() {
	const formContext = useContext( AdaptiveFormContext );
}
```

So there is a common pattern to avoid the redundancy above.

```js
import { useAdaptiveFormContext } from '.~/components/adaptive-form';

function InputComponent() {
	const formContext = useAdaptiveFormContext();
}
```

Comparing these two uses, should the following one be less redundant?

> Why not to have a context folder with the contexts (this one for now) and just call useContext() in the consumer side?

There are usually very few situations where React Context will be used, so I supposed there is still no need for a dedicated directory to centralize the only context at present.